File: token_tracker.py
import os
import json
import time
import random
import requests
from datetime import datetime

# API Configuration
url = "https://mainnet.helius-rpc.com/?api-key="  # API key from Helius

# Token configuration
TOKEN_MINT = "6rQt2W1kPAZGhuSMJmPnRgJndLb1HfK6E4cHiQTxoxzq"
TOKEN_DECIMALS = 6  # Number of decimals in the token, check on Solscan if not sure

# Define holding tiers
TIERS = [
    {"id": 1, "name": "Tier 1", "min": 10000, "max": 99999},
    {"id": 2, "name": "Tier 2", "min": 100000, "max": 499999},
    {"id": 3, "name": "Tier 3", "min": 500000, "max": 999999},
    {"id": 4, "name": "Tier 4", "min": 1000000, "max": 4999999},
    {"id": 5, "name": "Tier 5", "min": 5000000, "max": float('inf')}
]

# Define winner count per tier
WINNERS_BY_TIER = {
    1: 5,  # Tier 1: 5 winners
    2: 4,  # Tier 2: 4 winners
    3: 3,  # Tier 3: 3 winners
    4: 2,  # Tier 4: 2 winners
    5: 1   # Tier 5: 1 winner
}

# Time interval in seconds (10 minutes)
INTERVAL_SECONDS = 10 * 60

# Create logs directory if it doesn't exist
log_dir = './logs'
if not os.path.exists(log_dir):
    os.makedirs(log_dir)


def get_tier_for_amount(amount):
    """
    Get the tier for a given amount
    
    Args:
        amount (float): Token amount
        
    Returns:
        dict or None: Tier object or None if no tier matches
    """
    for tier in TIERS:
        if tier["min"] <= amount <= tier["max"]:
            return tier
    return None  # No tier matches (amount is less than the minimum of any tier)


def get_formatted_datetime():
    """
    Format current date and time
    
    Returns:
        str: Formatted date and time
    """
    return datetime.now().strftime('%Y-%m-%d %H:%M:%S')


def log_to_file(message):
    """
    Append a message to the log file
    
    Args:
        message (str): Message to log
    """
    log_file = f"{log_dir}/tracker_{datetime.now().strftime('%Y-%m-%d')}.log"
    with open(log_file, 'a') as f:
        f.write(f"{get_formatted_datetime()} - {message}\n")


async def find_holders():
    """
    Find all token holders and categorize them by tier
    
    Returns:
        list: Array of wallet data objects
    """
    print('Scanning for token holders...')
    log_to_file('Starting token holder scan')
    
    page = 1
    all_owners = {}  # Dictionary to store owner-amount pairs
    
    # Track counts for each tier
    tier_counts = {tier["id"]: 0 for tier in TIERS}
    
    total_wallets = 0
    wallets_below_min_tier = 0

    try:
        while True:
            print(f"Fetching page {page}...")
            
            payload = {
                "jsonrpc": "2.0",
                "method": "getTokenAccounts",
                "id": "helius-test",
                "params": {
                    "page": page,
                    "limit": 1000,
                    "displayOptions": {},
                    "mint": TOKEN_MINT,
                }
            }
            
            response = requests.post(url, headers={"Content-Type": "application/json"}, json=payload)

            # Check if any error in the response
            if not response.ok:
                print(f"Error: {response.status_code}, {response.reason}")
                log_to_file(f"API Error: {response.status_code}, {response.reason}")
                break

            data = response.json()

            if not data.get("result") or not data["result"].get("token_accounts") or len(data["result"]["token_accounts"]) == 0:
                print(f"No more results. Total pages: {page - 1}")
                break
            
            print(f"Processing results from page {page}")
            
            # Store both owner address and token amount
            for account in data["result"]["token_accounts"]:
                # Get the amount from the account
                raw_amount = float(account.get("amount", 0))
                
                # Convert to human-readable format by dividing by 10^TOKEN_DECIMALS
                amount = raw_amount / (10 ** TOKEN_DECIMALS)
                
                # If owner already exists in dict, add to their total balance
                if account["owner"] in all_owners:
                    all_owners[account["owner"]] += amount
                else:
                    all_owners[account["owner"]] = amount
            
            page += 1

        # Convert dict to list of objects with wallet, amount, and tier information
        wallet_data = []
        for wallet, amount in all_owners.items():
            tier = get_tier_for_amount(amount)
            if tier:
                tier_counts[tier["id"]] += 1
            else:
                wallets_below_min_tier += 1
            total_wallets += 1
            
            wallet_data.append({
                "wallet": wallet,
                "amount": f"{amount:.5f}",  # Format amount to 5 decimal places for readability
                "tier": tier["id"] if tier else 0
            })

        # Sort wallets by amount in descending order (highest holders first)
        wallet_data.sort(key=lambda x: float(x["amount"]), reverse=True)

        # Output wallet data
        with open("output.json", "w") as f:
            json.dump(wallet_data, f, indent=2)
        
        # Create tier summary
        tier_summary = {
            "totalWallets": total_wallets,
            "walletsBelowMinTier": wallets_below_min_tier,
            "tiers": [{
                **tier,
                "walletCount": tier_counts[tier["id"]],
                "percentage": f"{(tier_counts[tier['id']] / total_wallets * 100):.2f}%"
            } for tier in TIERS]
        }
        
        # Output tier summary
        with open("tier_summary.json", "w") as f:
            json.dump(tier_summary, f, indent=2)

        print(f"Finished scanning token mint: {TOKEN_MINT}")
        print(f"Total holders: {len(all_owners)}")
        print("Tier summary:")
        print(f"Below all tiers (<{TIERS[0]['min']}): {wallets_below_min_tier} wallets ({wallets_below_min_tier / total_wallets * 100:.2f}%)")
        
        for tier in TIERS:
            print(f"{tier['name']} ({tier['min']}-{'MAX' if tier['max'] == float('inf') else tier['max']}): {tier_counts[tier['id']]} wallets ({tier_counts[tier['id']] / total_wallets * 100:.2f}%)")

        log_to_file(f"Finished token scan. Found {len(all_owners)} holders")
        
        return wallet_data
    except Exception as error:
        print(f"Error scanning token holders: {error}")
        log_to_file(f"Error scanning token holders: {error}")
        return []


def select_winners(holders, winner_count_per_tier=None):
    """
    Select random winners from each tier
    
    Args:
        holders (list): List of wallet data objects
        winner_count_per_tier (dict, optional): Dict with tier IDs as keys and winner counts as values
            
    Returns:
        dict: Dictionary with tier IDs as keys and lists of winners as values
    """
    if winner_count_per_tier is None:
        winner_count_per_tier = WINNERS_BY_TIER
        
    print('Selecting winners...')
    log_to_file('Selecting winners')
    
    winners = {}
    
    # Group holders by tier
    holders_by_tier = {}
    for holder in holders:
        tier = holder.get("tier", 0)
        if tier == 0:  # Skip wallets that don't belong to any tier
            continue
            
        if tier not in holders_by_tier:
            holders_by_tier[tier] = []
        
        holders_by_tier[tier].append(holder)
    
    # For each tier, select random winners
    for tier_id, tier_holders in holders_by_tier.items():
        if tier_id not in winner_count_per_tier:
            continue
            
        count = winner_count_per_tier[tier_id]
        
        # If we have fewer holders than the winner count, select all holders
        if len(tier_holders) <= count:
            winners[tier_id] = tier_holders
            print(f"Tier {tier_id}: Only {len(tier_holders)} holders, selected all as winners")
            log_to_file(f"Tier {tier_id}: Only {len(tier_holders)} holders, selected all as winners")
            
            # Log each winner's address
            for winner in tier_holders:
                log_to_file(f"Tier {tier_id} Winner: {winner['wallet']} with {winner['amount']} tokens")
        else:
            # Randomly select winners
            tier_winners = random.sample(tier_holders, count)
            winners[tier_id] = tier_winners
            print(f"Tier {tier_id}: Selected {count} winners out of {len(tier_holders)} holders")
            log_to_file(f"Tier {tier_id}: Selected {count} winners out of {len(tier_holders)} holders")
            
            # Log each winner's address
            for winner in tier_winners:
                log_to_file(f"Tier {tier_id} Winner: {winner['wallet']} with {winner['amount']} tokens")
    
    # Save winners to file
    with open("winners.json", "w") as f:
        # Convert to a serializable format
        serializable_winners = {
            tier_id: [winner for winner in tier_winners]
            for tier_id, tier_winners in winners.items()
        }
        json.dump(serializable_winners, f, indent=2)
    
    return winners


def generate_summary(winners):
    """
    Generate a summary of winners
    
    Args:
        winners (dict): Dictionary with tier IDs as keys and lists of winners as values
            
    Returns:
        dict: Summary object
    """
    print('Generating summary...')
    
    # Create summary object
    summary = {
        "timestamp": get_formatted_datetime(),
        "totalWinners": sum(len(tier_winners) for tier_winners in winners.values()),
        "tiers": []
    }
    
    # Add tier information to summary
    for tier in TIERS:
        tier_id = tier["id"]
        tier_winners = winners.get(tier_id, [])
        
        tier_summary = {
            "id": tier_id,
            "name": tier["name"],
            "winnerCount": len(tier_winners),
            "totalTokens": sum(float(winner["amount"]) for winner in tier_winners)
        }
        
        summary["tiers"].append(tier_summary)
    
    # Save summary to file
    # with open("winners_summary.json", "w") as f:
    #     json.dump(summary, f, indent=2)
    
    return summary


async def run_token_tracker():
    """Run the complete token tracking sequence"""
    start_time = time.time()
    print(f"\n----- Token Tracker Run: {get_formatted_datetime()} -----")
    log_to_file("Starting token tracker run")
    
    try:
        # Find all token holders
        holders = await find_holders()
        
        if not holders:
            print("No holders found, stopping tracking sequence")
            log_to_file("No holders found, stopping tracking sequence")
            return
        
        # Select winners
        winners = select_winners(holders)
        
        # Generate summary
        summary = generate_summary(winners)
        
        # Print summary
        print("\nWinner Selection Summary:")
        print(f"Total Winners: {summary['totalWinners']}")
        print("Tier Breakdown:")
        
        for tier_summary in summary["tiers"]:
            print(f"{tier_summary['name']}: {tier_summary['winnerCount']} winners, {tier_summary['totalTokens']:.2f} tokens total")
        
        # Print winners with their addresses and tiers (similar to JS version)
        print("\nSelected Winners:")
        for tier_id, tier_winners in winners.items():
            print(f"\n--- Tier {tier_id} Winners ---")
            for winner in tier_winners:
                print(f"Address: {winner['wallet']}, Amount: {winner['amount']}")
                log_to_file(f"Winner: Tier {tier_id}, Address: {winner['wallet']}, Amount: {winner['amount']}")
        
        # Calculate runtime
        run_time = time.time() - start_time
        print(f"\nToken tracking complete in {run_time:.2f} seconds")
        log_to_file(f"Token tracking complete in {run_time:.2f} seconds")
        
        # Record the run timestamp (similar to JS version)
        timestamp_file = './last_run.json'
        with open(timestamp_file, 'w') as f:
            json.dump({
                "lastRunAt": datetime.now().isoformat(),
                "completedAt": datetime.now().isoformat(),
                "durationSeconds": run_time,
                "nextRunAt": (datetime.now().timestamp() + INTERVAL_SECONDS)
            }, f, indent=2)
        
    except Exception as error:
        print(f"Error in token tracking sequence: {error}")
        log_to_file(f"Error in token tracking sequence: {error}")


async def main():
    """Main function that runs the token tracker continuously"""
    print(f"Token Tracker starting up at {get_formatted_datetime()}")
    log_to_file("Token Tracker starting up")
    
    while True:
        try:
            await run_token_tracker()
        except Exception as error:
            print(f"Unhandled error in token tracker: {error}")
            log_to_file(f"Unhandled error in token tracker: {error}")
        
        print(f"Waiting {INTERVAL_SECONDS} seconds for next run...")
        time.sleep(INTERVAL_SECONDS)


if __name__ == "__main__":
    import asyncio
    
    try:
        # Add hook to handle keyboard interrupts gracefully
        loop = asyncio.get_event_loop()
        loop.run_until_complete(main())
    except KeyboardInterrupt:
        print("\nToken Tracker shutting down...")
        log_to_file("Token Tracker shutting down by user request")
    except Exception as e:
        print(f"Unhandled error in main process: {e}")
        log_to_file(f"Unhandled error in main process: {e}")

File: token_tracker.js
const fs = require("fs");
const url = `https://mainnet.helius-rpc.com/?api-key=`; // API key from Helius

// Token configuration
const TOKEN_MINT = "6rQt2W1kPAZGhuSMJmPnRgJndLb1HfK6E4cHiQTxoxzq";
const TOKEN_DECIMALS = 6; // Number of decimals in the token check on Solscan if not sure

// Define holding tiers
const TIERS = [
  { id: 1, name: "Tier 1", min: 10000, max: 99999 },
  { id: 2, name: "Tier 2", min: 100000, max: 499999 },
  { id: 3, name: "Tier 3", min: 500000, max: 999999 },
  { id: 4, name: "Tier 4", min: 1000000, max: 4999999 },
  { id: 5, name: "Tier 5", min: 5000000, max: Number.MAX_SAFE_INTEGER }
];

// Define winner count per tier
const WINNERS_BY_TIER = {
  1: 5,  // Tier 1: 5 winners
  2: 4,  // Tier 2: 4 winners
  3: 3,  // Tier 3: 3 winners
  4: 2,  // Tier 4: 2 winners
  5: 1   // Tier 5: 1 winner
};

// Time interval in milliseconds (10 minutes)
const INTERVAL_MS = 10 * 60 * 1000;

// Create logs directory if it doesn't exist
const logDir = './logs';
if (!fs.existsSync(logDir)) {
  fs.mkdirSync(logDir);
}

/**
 * Get the tier for a given amount
 * @param {number} amount - Token amount
 * @returns {object|null} - Tier object or null if no tier matches
 */
const getTierForAmount = (amount) => {
  for (const tier of TIERS) {
    if (amount >= tier.min && amount <= tier.max) {
      return tier;
    }
  }
  return null; // No tier matches (amount is less than the minimum of any tier)
};

/**
 * Format current date and time
 * @returns {string} Formatted date and time
 */
const getFormattedDateTime = () => {
  const now = new Date();
  return now.toISOString().replace(/T/, ' ').replace(/\..+/, '');
};

/**
 * Append a message to the log file
 * @param {string} message - Message to log
 */
const logToFile = (message) => {
  const logFile = `${logDir}/tracker_${new Date().toISOString().split('T')[0]}.log`;
  fs.appendFileSync(logFile, `${getFormattedDateTime()} - ${message}\n`);
};

/**
 * Find all token holders and categorize them by tier
 * @returns {Array} Array of wallet data objects
 */
const findHolders = async () => {
  console.log('Scanning for token holders...');
  logToFile('Starting token holder scan');
  
  let page = 1;
  let allOwners = new Map(); // Map to store owner-amount pairs
  
  // Track counts for each tier
  const tierCounts = TIERS.reduce((acc, tier) => {
    acc[tier.id] = 0;
    return acc;
  }, {});
  
  let totalWallets = 0;
  let walletsBelowMinTier = 0;

  try {
    while (true) {
      console.log(`Fetching page ${page}...`);
      
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          method: "getTokenAccounts",
          id: "helius-test",
          params: {
            page: page,
            limit: 1000,
            displayOptions: {},
            mint: TOKEN_MINT,
          },
        }),
      });

      // Check if any error in the response
      if (!response.ok) {
        console.log(`Error: ${response.status}, ${response.statusText}`);
        logToFile(`API Error: ${response.status}, ${response.statusText}`);
        break;
      }

      const data = await response.json();

      if (!data.result || data.result.token_accounts.length === 0) {
        console.log(`No more results. Total pages: ${page - 1}`);
        break;
      }
      
      console.log(`Processing results from page ${page}`);
      
      // Store both owner address and token amount
      data.result.token_accounts.forEach((account) => {
        // Get the amount from the account
        const rawAmount = account.amount ? Number(account.amount) : 0;
        
        // Convert to human-readable format by dividing by 10^TOKEN_DECIMALS
        const amount = rawAmount / Math.pow(10, TOKEN_DECIMALS);
        
        // If owner already exists in map, add to their total balance
        if (allOwners.has(account.owner)) {
          const currentAmount = allOwners.get(account.owner);
          allOwners.set(account.owner, currentAmount + amount);
        } else {
          allOwners.set(account.owner, amount);
        }
      });
      
      page++;
    }

    // Convert Map to array of objects with wallet, amount, and tier information
    const walletData = Array.from(allOwners.entries()).map(([wallet, amount]) => {
      const tier = getTierForAmount(amount);
      if (tier) {
        tierCounts[tier.id]++;
      } else {
        walletsBelowMinTier++;
      }
      totalWallets++;
      
      return {
        wallet,
        amount: amount.toFixed(5), // Format amount to 5 decimal places for readability
        tier: tier ? tier.id : 0
      };
    });

    // Sort wallets by amount in descending order (highest holders first)
    walletData.sort((a, b) => parseFloat(b.amount) - parseFloat(a.amount));

    // Output wallet data
    fs.writeFileSync(
      "output.json",
      JSON.stringify(walletData, null, 2)
    );
    
    // Create tier summary
    const tierSummary = {
      totalWallets,
      walletsBelowMinTier,
      tiers: TIERS.map(tier => ({
        ...tier,
        walletCount: tierCounts[tier.id],
        percentage: ((tierCounts[tier.id] / totalWallets) * 100).toFixed(2) + '%'
      }))
    };
    
    // Output tier summary
    fs.writeFileSync(
      "tier_summary.json",
      JSON.stringify(tierSummary, null, 2)
    );

    console.log(`Finished scanning token mint: ${TOKEN_MINT}`);
    console.log(`Total holders: ${allOwners.size}`);
    console.log("Tier summary:");
    console.log(`Below all tiers (<${TIERS[0].min}): ${walletsBelowMinTier} wallets (${((walletsBelowMinTier / totalWallets) * 100).toFixed(2)}%)`);
    
    TIERS.forEach(tier => {
      console.log(`${tier.name} (${tier.min}-${tier.max === Number.MAX_SAFE_INTEGER ? 'MAX' : tier.max}): ${tierCounts[tier.id]} wallets (${((tierCounts[tier.id] / totalWallets) * 100).toFixed(2)}%)`);
    });

    logToFile(`Finished token scan. Found ${allOwners.size} holders`);
    
    return walletData;
  } catch (error) {
    console.error('Error scanning token holders:', error);
    logToFile(`Error scanning token holders: ${error.message}`);
    return [];
  }
};

/**
 * Select random winners from each tier
 * @param {Array} holders - Array of wallet data objects
 * @param {Object} winnerCountPerTier - Object with tier IDs as keys and winner counts as values
 * @returns {Object} Object with tier IDs as keys and arrays of winners as values
 */
const selectWinners = (holders, winnerCountPerTier = WINNERS_BY_TIER) => {
  console.log('Selecting winners...');
  logToFile('Selecting winners');
  
  const tierHolders = {
    1: [],
    2: [],
    3: [],
    4: [],
    5: []
  };
  
  // Group holders by tier
  holders.forEach(holder => {
    if (holder.tier > 0) {
      tierHolders[holder.tier].push(holder);
    }
  });
  
  const winners = {};
  
  // Select winners for each tier
  for (const tier in tierHolders) {
    winners[tier] = [];
    
    if (tierHolders[tier].length === 0) {
      console.log(`No holders found in Tier ${tier}`);
      continue;
    }
    
    // Select random winners without replacement
    const selectedIndices = new Set();
    const tierHolderCount = tierHolders[tier].length;
    const winnerCount = winnerCountPerTier[tier] || 1;
    const actualWinnerCount = Math.min(winnerCount, tierHolderCount);
    
    while (selectedIndices.size < actualWinnerCount) {
      const randomIndex = Math.floor(Math.random() * tierHolderCount);
      if (!selectedIndices.has(randomIndex)) {
        selectedIndices.add(randomIndex);
        const winner = tierHolders[tier][randomIndex];
        
        winners[tier].push({
          wallet: winner.wallet,
          amount: winner.amount,
          tier: tier
        });
      }
    }
  }
  
  return winners;
};

/**
 * Generate a summary of winners
 * @param {Object} winners - Object with tier IDs as keys and arrays of winners as values
 * @returns {Object} Summary object
 */
const generateSummary = (winners) => {
  let totalWinners = 0;
  let summary = {};
  
  for (const tier in winners) {
    totalWinners += winners[tier].length;
    
    summary[tier] = {
      winnerCount: winners[tier].length
    };
  }
  
  return {
    totalWallets: totalWinners,
    tierSummary: summary
  };
};

/**
 * Run the complete token tracking sequence
 */
const runTokenTracker = async () => {
  const startTime = new Date();
  console.log(`\n--- Starting token tracker at ${getFormattedDateTime()} ---`);
  logToFile('--- Starting token tracker sequence ---');
  
  try {
    // Step 1: Find all token holders
    const holders = await findHolders();
    
    if (holders.length === 0) {
      console.log('No holders found. Aborting winner selection.');
      logToFile('No holders found. Aborting winner selection.');
      return;
    }
    
    // Step 2: Select winners from each tier
    const allWinners = selectWinners(holders, WINNERS_BY_TIER);
    
    // Step 3: Generate and save summary
    const summary = generateSummary(allWinners);
    
    // Save results to file
    fs.writeFileSync(
      './winners.json', 
      JSON.stringify(allWinners, null, 2), 
      'utf8'
    );
    
    console.log('\nWinners selected successfully!');
    console.log(`Total winners: ${summary.totalWallets}`);
    
    // Print winners with their addresses and tiers
    for (const tier in allWinners) {
      console.log(`\n--- Tier ${tier} Winners ---`);
      allWinners[tier].forEach(winner => {
        console.log(`Address: ${winner.wallet}, Tier: ${winner.tier}`);
      });
    }
    
    console.log('\nDetails saved to winners.json');
    
    // Record the run timestamp
    const endTime = new Date();
    const durationMs = endTime - startTime;
    const timestampFile = './last_run.json';
    fs.writeFileSync(timestampFile, JSON.stringify({ 
      lastRunAt: startTime.toISOString(),
      completedAt: endTime.toISOString(),
      durationMs: durationMs,
      nextRunAt: new Date(Date.now() + INTERVAL_MS).toISOString()
    }, null, 2));
    
    logToFile(`Completed token tracker sequence. Duration: ${durationMs}ms`);
    console.log(`\n--- Completed token tracker at ${getFormattedDateTime()} ---`);
    console.log(`Duration: ${(durationMs / 1000).toFixed(2)} seconds`);
    console.log(`Next run scheduled at: ${new Date(Date.now() + INTERVAL_MS).toLocaleString()}`);
    
  } catch (error) {
    console.error('Error in token tracker sequence:', error);
    logToFile(`Error in token tracker sequence: ${error.message}`);
  }
};

/**
 * Main function that runs the token tracker continuously
 */
const main = async () => {
  // Run immediately on startup
  await runTokenTracker();
  
  // Schedule to run at the interval
  console.log(`\nToken tracker will run every ${INTERVAL_MS / 60000} minutes.`);
  setInterval(runTokenTracker, INTERVAL_MS);
};

// Start the token tracker
main().catch(err => {
  console.error('Unhandled error in main process:', err);
  logToFile(`Unhandled error in main process: ${err.message}`);
});

// Keep the process alive
process.stdin.resume();

// Handle signals gracefully
process.on('SIGINT', () => {
  console.log('\nShutting down token tracker...');
  logToFile('Token tracker was shutdown by user (SIGINT)');
  process.exit(0);
});

process.on('SIGTERM', () => {
  console.log('\nShutting down token tracker...');
  logToFile('Token tracker was shutdown (SIGTERM)');
  process.exit(0);
});
